cli/src/commands/extensions/api.rs

```diff
@@ -0,0 +1,254 @@
+use std::cell::RefCell;
+use std::ops::Deref;
+use std::path::Path;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::str::FromStr;
+
+use anyhow::{anyhow, Context, Error, Result};
+use deno_core::{op, OpDecl, OpState};
+use futures::future::BoxFuture;
+
+use phylum_types::types::auth::{AccessToken, RefreshToken};
+use phylum_types::types::common::{JobId, ProjectId};
+use phylum_types::types::job::JobStatusResponse;
+use phylum_types::types::package::{
+    Package, PackageDescriptor, PackageStatusExtended, PackageType,
+};
+use phylum_types::types::project::ProjectDetailsResponse;
+
+use crate::commands::parse::{get_packages_from_lockfile, LOCKFILE_PARSERS};
+use crate::config::get_current_project;
+use crate::{api::PhylumApi, auth::UserInfo};
+
+/// Holds either an unawaited, boxed `Future`, or the result of awaiting the future.
+pub(crate) enum OnceFuture<T: Unpin> {
+    Future(BoxFuture<'static, T>),
+    Awaited(T),
+}
+
+impl<T: Unpin> OnceFuture<T> {
+    fn new(inner: BoxFuture<'static, T>) -> Self {
+        OnceFuture::Future(inner)
+    }
+
+    async fn get(&mut self) -> &T {
+        match *self {
+            OnceFuture::Future(ref mut inner) => {
+                *self = OnceFuture::Awaited(inner.await);
+                match *self {
+                    OnceFuture::Future(..) => unreachable!(),
+                    OnceFuture::Awaited(ref mut inner) => inner,
+                }
+            }
+            OnceFuture::Awaited(ref mut inner) => inner,
+        }
+    }
+}
+
+/// Opaquely encapsulates the extension state.
+pub struct ExtensionState(OnceFuture<Result<Rc<PhylumApi>>>);
+
+impl From<BoxFuture<'static, Result<PhylumApi>>> for ExtensionState {
+    fn from(t: BoxFuture<'static, Result<PhylumApi>>) -> Self {
+        Self(OnceFuture::new(Box::pin(async { t.await.map(Rc::new) })))
+    }
+}
+
+/// Wraps a shared, counted reference to the `PhylumApi` object. The reference can be safely
+/// extracted from `Rc<RefCell<OpState>>` and cloned; it will not require mutable access to the
+/// owning `RefCell`, so the mutable borrow to it may be dropped.
```

Comment:
```suggestion
/// Wraps a shared, counted reference to the `PhylumApi` object.
///
/// The reference can be safely extracted from `Rc<RefCell<OpState>>` and cloned; it will not require mutable access to the owning `RefCell`, so the mutable borrow to it may be dropped.
```